test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py

```diff
@@ -417,6 +417,8 @@ def author(value):
         value = [value]
 
     for line in value:
+        if line is None:
+            raise Invalid("Specify an author")
```

Comment:
I think it's better to have `if not isinstance(line, string_types): continue`. That way all crashes are prevented, and not just the special case when `author` is `None` or a list containing `None`.

The schema itself already checks whether `author` is either `None`, a string, or a list of strings, and will report everything that is none of these.

Reply:
I agree with checking against string types. That was my original direction.

But I don't think the schema check is working correctly without a patch of some sort here.

Reply:
The schema check works correctly. (Which isn't necessarily the way you expect it to work :) ) The problem is the `author` function, which crashes for some inputs. It must not crash, and either return normally or raise a proper error like `Invalid`.

Reply:
Sorry, my eyes kind of glossed over the `continue` in your suggestion.

Testing that... it doesn't seem to work. In the updated PR now it gives

```
Running sanity test 'validate-modules' with Python 3.7
ERROR: Found 2 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/redshift.py:0:0: invalid-documentation: DOCUMENTATION.author: Specify author as a list of strings for dictionary value @ data['author']. Got None
ERROR: plugins/modules/redshift_subnet_group.py:0:0: invalid-documentation: DOCUMENTATION.author: Specify author as a list of strings for dictionary value @ data['author']. Got None
See documentation for help: https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html
```

But if I use the `continue` where the `raise` is now, there are no `validate-modules` failures, and no mention of the problematic `redshift.py` module.

I don't know why it's not giving the schema failure, in this case it's that `"author": None`, not `"author": [None]`, in the actual DOCUMENTATION string.

Reply:
`None` is a valid value for `author` though, according to the schema. If the schema is correct, no error should be reported for the redshift module - or at least not for `author`.

Once `version_added` is checked for collections, it would have complained about that one, since `"1.0.0 - \"Jens Carl (@j-carl), Hothead Games Inc.\" - \"Rafael Driutti (@rafaeldriutti)\""` is NOT a valid version number.

Reply:
> None is a valid value for author though, according to the schema. If the schema is correct, no error should be reported for the redshift module - or at least not for author.

I don't wish to take a position on this. The only thing I'm sure about is that it shouldn't give that TypeError in this case which can only be debugged mucking with the Ansible source code itself. Whether the new behavior is pass or fail... I only had a slight leaning toward fail.

I don't know who the ideal referee is for `ansible-test` is, but if I did I would `@` them here so I can defer to their judgement.

Reply:
I don't know either. Usually such questions are discussed in the public project meeting, if nobody from core team decides before that happens (looking at @sivel :) ).

In any case, if an empty author is not acceptable, you should better fix the schema and don't do this error reporting in the `author` function, or you change the schema to just run this function and not do other checks, and do better error reporting (right now there's only a generic `Specify author as a string or a list of strings, or not specify author at all` error which doesn't tell what it actually encountered, only that it didn't like the input).

Reply:
Since the existing voluptuous schema defines `None` as a valid option, we should probably make the `author` function adhere to that as well.  However the schema is tricky.  `author: null` is allowed, however `author: [null]` is not.

I think we might need something like this:

```diff
diff --git a/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py b/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
index 22b1a53d11..9587b29c98 100644
--- a/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
+++ b/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py
@@ -412,15 +412,21 @@ def deprecation_schema(for_collection):
 
 
 def author(value):
+    if value is None:
+        return value
 
     if not is_iterable(value):
         value = [value]
 
     for line in value:
+        if not isinstance(line, string_types):
+            continue
         m = author_line.search(line)
         if not m:
             raise Invalid("Invalid author")
 
+    return value
+
 
 def doc_schema(module_name, for_collection=False, deprecated_module=False):
 ```

Reply:
I committed that, but it does change this case of redshift into a pass, not a failure.

Reply:
The redshift module says `author: null` (where `null` is implicit), so it is correct with regard to satisfying the schema for `author`. Sooner or later we will have a check for `version_added` in collections, and that check would flag this case (since here `version_added` is garbled).

Reply:
Yep, as of now this allows `author` to pass, since it actually does satisfy the schema.  The real issue in `community.aws` was the `version_added` was placed in the wrong spot, which actually invalidates `version_added`.

Unfortunately we don't validate `version_added` in collections currently, because at the time there had been no standard, and that is still being worked out to some extent.

Reply:
That will come soonish, I hope. Whenever all more pressing things are done and I find time to work on my version_added PR again ;)

Reply:
This PR adds `version_added` validation: #70869. It was already there for module options and return values, but not for the top-level `version_added` of a module in a collection.